core/src/main/java/jenkins/security/apitoken/ApiTokenStore.java

```diff
@@ -0,0 +1,381 @@
+        // 16x8=128bit worth of randomness, using brute-force you need on average 2^127 tries (~10^37)
+        byte[] random = new byte[16];
+        RANDOM.nextBytes(random);
+        String secretValue = Util.toHexString(random);
+        String tokenTheUserWillUse = HASH_VERSION + secretValue;
+        assert tokenTheUserWillUse.length() == 2 + 32;
+        
+        String secretValueHashed = this.plainSecretToHashInHex(secretValue);
+        
+        HashValue hashValue = new HashValue(HASH_VERSION, secretValueHashed);
+        HashedToken token = HashedToken.buildNew(name, hashValue);
+        
+        this.addToken(token);
+        
+        return new TokenIdAndPlainValue(token.uuid, tokenTheUserWillUse);
+    }
+    
+    @SuppressFBWarnings("NP_NONNULL_RETURN_VIOLATION")
+    private @Nonnull String plainSecretToHashInHex(@Nonnull String secretValueInPlainText) {
+        byte[] hashBytes = plainSecretToHashBytes(secretValueInPlainText);
+        return Util.toHexString(hashBytes);
+    }
+    
+    private @Nonnull byte[] plainSecretToHashBytes(@Nonnull String secretValueInPlainText) {
+        // ascii is sufficient for hex-format
+        return hashedBytes(secretValueInPlainText.getBytes(StandardCharsets.US_ASCII));
+    }
+    
+    private @Nonnull byte[] hashedBytes(byte[] tokenBytes) {
+        MessageDigest digest;
+        try {
+            digest = MessageDigest.getInstance(HASH_ALGORITHM);
+        } catch (NoSuchAlgorithmException e) {
+            throw new AssertionError("There is no " + HASH_ALGORITHM + " available in this system");
+        }
+        return digest.digest(tokenBytes);
+    }
+    
+    public synchronized @CheckForNull HashedToken findMatchingToken(@Nonnull String token) {
+        String plainToken;
+        if (isLegacyToken(token)) {
+            plainToken = token;
+        } else {
+            plainToken = getHashOfToken(token);
+        }
+        
+        return searchMatch(plainToken);
+    }
+    
+    private boolean isLegacyToken(@Nonnull String token) {
+        return token.length() != TOKEN_LENGTH_V2;
+    }
+    
+    /**
+     * [2: version][32: real token]
+     * ^^^^^^^^^^^^----------------
+     */
+    private @Nonnull String getVersionOfToken(@Nonnull String token) {
+        return token.substring(0, 2);
+    }
+    
+    /**
+     * [2: version][32: real token]
+     * ------------^^^^^^^^^^^^^^^^
+     */
+    private @Nonnull String getHashOfToken(@Nonnull String token) {
+        return token.substring(2);
+    }
+    
+    private @CheckForNull HashedToken searchMatch(@Nonnull String plainToken) {
+        byte[] hashedBytes = plainSecretToHashBytes(plainToken);
+        for (HashedToken token : tokenList) {
+            if (token.match(hashedBytes)) {
+                return token;
+            }
+        }
+        
+        return null;
+    }
+    
+    public synchronized @CheckForNull HashedToken revokeToken(@Nonnull String tokenUuid) {
```

Comment:
in general would like to see much more javadoc.
why might this return null for example (without having to read the code)?